### DISCUSSION TITLE
Heterogeneous Graph with no features

### DISCUSSION THREAD

**Discussion by Zytrus** [pyg-team/pytorch_geometric#4408] [Q&A] [answered]
Hello,
I am new to GNNs and am currently working on creating a Heterogeneous GNN on graph-level predictions. I've seen many other discussions as well, but I didn't quite come across the same problem as I currently have.
My problem is that I have a graph with 3 node types and 5 edge types without features on any of them. I'd like to have features that consider the structural information and the types into consideration:

```py
data = HeteroData()

data['process'].x = ...
data['socket'].x = ...
data['file'].x = ...

data['process','is_parent_of','process'].edge_index = ...
data['process','writes','file'].edge_index = ...
data['file','read_by','process'].edge_index = ...
data['process','connects_to', 'socket'].edge_index = ...
data['socket','sends_to','process'].edge_index = ...
```
I have no features for the nodes in my dataset. The only feature I could give them is either a 1 for each, or the node degree, but as I've read in many papers the node degree is inefficient especially for heterogeneous graphs, because they're usually very similar and don't differentiate between different kinds of edges.
So my question would be how to give them features?

Giving features with pre-processing sounds reasonable, but how can you apply pre-processing, when the dataset is not initialized? 
A little help would be very much appreciated!

Another question is how batching the data works, so that each batch represents a graph in HeteroData(). Do you have to write a batch function, or how does batch() know, where the edges of Graph1 ends and Graph2 starts?

Best regards, Zytrus

**Comment by rusty1s** [comment, marked as answer]
This is tricky. In general, it's hard to get any better than using node degree statistics :)

You can also use `MetaPath2Vec` but this does not really work in an inductive learning scenario where you want to apply your model on unseen graphs. An additional alternative is to lift `transforms.LocalDegreeProfile` to the heterogeneous graph case.

For mini-batching, it works equivalent as in a homogeneous scenario, except that now you have a `batch` vector for each node type. You can then use `global_pooling` operators for each node type separately.

**Reply by Zytrus** [reply]
Hello rusty1s!
Thank you for the quick reply, so MetaPath2Vec would not work on the test dataset if I understood correctly? Since we'd be training with the training set and the test set would be a collection of graphs with a different structure than in the training set?

Also if there aren't any better ones than node degrees, then I'll just settle with node degrees, thank your help!

Best regards, Zytrus

**Reply by rusty1s** [reply]
Yeah, the embeddings produced by `MetaPath2Vec` might not share any similarities across different splits or even across different examples. I don't think it is suitable in your case. Let me know if making use of node degrees works for you :)

**Reply by Zytrus** [reply]
Alright thank you, would it be different if I preprocessed the whole dataset with MetaPath2Vec? Then unseen graphs from other datasets would not work, but technically the test set inside the dataset would have features then as well right?

**Reply by rusty1s** [reply]
This is not possible as well, as the embeddings from `MetaPath2Vec` across different graphs do not need to share any similarities. `MetaPath2Vec` is purely transductive.

**Reply by Zytrus** [reply]
Hello again rusty1s! (: 
I have the choice between node degree features for each node or a node degree specified by neighbor type, e.g.
A->B
A->C
A has node degree 2, but the specified feature would be [0,1,1] assuming we have 3 types. The first row represents connections to neighbor type A, the seoncd B and third C.

So my question would be, is there something special in the GNN I have to customize or change for this kind of feature?
I'm sorry if this seems like an amateur question, but I'm not sure depending on what feature I have, if there needs to be adjustments in the model or if the model will just learn with it automatically...

Looking forward to a reply!

**Reply by rusty1s** [reply]
I don't think you need to make any adjustments for this on the GNN model design. I would probably go with per-node type degree features, and simply concatenate them together.

**Reply by Zytrus** [reply]
Concatenate together? Is that different than using a vector as a feature?

**Reply by rusty1s** [reply]
No, that means the same :)

**Reply by Zytrus** [reply]
Thank you very much for the help so far! (:
For now I am using the node degrees, but will later go for the node degrees with specialized type.
But currently an Exception appears, and I don't really know how it's caused:

```
    super(SystemCallGraphs, self).__init__(root, transform, pre_transform, pre_filter)
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\in_memory_dataset.py", line 57, in __init__
    super().__init__(root, transform, pre_transform, pre_filter)
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\dataset.py", line 88, in __init__
    self._process()
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\dataset.py", line 171, in _process
    self.process()
  File "C:\Users\chenx\Documents\Privacy\SystemCallGraph_node_degree.py", line 220, in process
    data, slices = self.collate(hetero_list)
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\in_memory_dataset.py", line 112, in collate
    data, slices, _ = collate(
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\collate.py", line 85, in collate
    value, slices, incs = _collate(attr, values, data_list, stores,
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\collate.py", line 131, in _collate
    slices = cumsum([value.size(cat_dim or 0) for value in values])
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\collate.py", line 131, in <listcomp>
    slices = cumsum([value.size(cat_dim or 0) for value in values])
TypeError: 'int' object is not callable
```

It looks like something went wrong, because I called self.collate on my HeteroData() array, do you have a guess what the cause could be?

**Reply by Zytrus** [reply]
I think I have a hunch what it could be. I have some graphs in which certain node types don't appear. Is that supported by PyTorch Geo?

**Reply by rusty1s** [reply]
This is not supported at the moment. You may need to add dummy node types for these graphs (e.g., by setting `data[node_type].x = torch.empty((0, num_features), dtype=torch.float)`).

**Reply by Zytrus** [reply]
This error appears when I have ```data[node_type].x = torch.empty((0, 1), dtype=torch.float)``` for the empty features:
```
  File "C:\Users\chenx\Documents\Privacy\SystemCallGraph_node_degree.py", line 244, in process        
    data, slices = self.collate(hetero_list)
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\in_memory_dataset.py", line 112, in collate
    data, slices, _ = collate(
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\collate.py", line 85, in collate
    value, slices, incs = _collate(attr, values, data_list, stores,
  File "C:\Users\chenx\miniconda3\lib\site-packages\torch_geometric\data\collate.py", line 147, in _collate
    value = torch.cat(values, dim=cat_dim or 0, out=out)
RuntimeError: Tensors must have same number of dimensions: got 1 and 2
```
And ```'int' object is not callable``` appears again when I have ```data[node_type].x = torch.empty( 0, dtype=torch.float )```. So now I have dummy values for each graph, do you have a hunch what this 'int' error could represent?

Thank you for the help so far!

Best regards, Zytrus

**Reply by Zytrus** [reply]
The error shows me this line of code:
```python
slices = cumsum([value.size(cat_dim or 0) for value in values])
```
Here value.size, which is an int is called like a function with (cat_dim or 0), could that be a bug?

**Reply by rusty1s** [reply]
IMO, this may only happen in case this attribute is a tensor in one example while it might be an int in another. Can you confirm that this is not the caee? Otherwise, please try to construct a simple reproducible example and I am happy to take a look.

**Reply by Zytrus** [reply]
This is where my mistake was thank you! I didn't know the 'Array type' had impact on the processing, I thought Tensor and Numpy would do the trick. Now it's running through thank you very much!
